### DISCUSSION TITLE
Stuck at implementation of middleware

### DISCUSSION THREAD

**Discussion by SiddheshKanawade** [tokio-rs/axum#1113] [Q&A] [answered]
I am trying to integrate Casbin with Axum to implement a access control middleware. I am using [this](https://docs.rs/axum/latest/axum/middleware/index.html#towerservice-and-pinboxdyn-future) template to implement the middleware. 
I am getting  the following error:

![Screenshot from 2022-06-22 17-27-36](https://user-images.githubusercontent.com/85673539/175023574-92fa4d8e-32f2-40d3-9771-39a589ae9024.png)

It seems I am not able to properly define Service, but not sure where I am going wrong

Following is the relevant code:
```
impl<S> Layer<S> for CasbinAxumService {
    type Service = CasbinAxumMiddleware<S>;

    fn layer(&self, service: S) -> Self::Service {
        CasbinAxumMiddleware {
            enforcer: self.enforcer.clone(),
            service: Rc::new(RefCell::new(service)),
        }
    }
}
```
```
#[derive(Clone)]
pub struct CasbinAxumMiddleware<S> {
    service: Rc<RefCell<S>>,
    enforcer: Arc<RwLock<CachedEnforcer>>,
}
```
```
impl<S> Service<Request<Body>> for CasbinAxumMiddleware<S>
where
    S: Service<Request<Body>, Response = Response> + Send + 'static,
    S::Future: Send + 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    // `BoxFuture` is a type alias for `Pin<Box<dyn Future + Send + 'a>>`
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(self.service.poll_ready(cx)))
    }

    fn call(&mut self, mut request: Request<Body>) -> Self::Future {
        let future = self.service.call(request);
        Box::pin(async move {
            let response: Response = future.await?;
            Ok(response)
        })
    }
}
```

**Comment by davidpdrsn** [comment, marked as answer]
You probably don't want to use `Rc<RefCell<S>>` inside a middleware. Why do you think you need that? Usually we recommend cloning the inner service.

Also if you just wanna write a middleware for your app and don't mind using something axum specific then I recommend https://docs.rs/axum/latest/axum/middleware/fn.from_fn.html. See https://docs.rs/axum/latest/axum/middleware/index.html#writing-middleware for the trade offs.
